### PR TITLE
iox-#1032 Refactor 'IOX_REQUIRE'

### PR DESCRIPTION
**Pull request by elBoberido**
## Pre-Review Checklist for the PR Author

1. [x] Add a second reviewer for complex new features or larger refactorings
1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

Currently `IOX_REQUIRE` can only be used with an error code which makes it semantically similar to `IOX_REPORT_FATAL_IF`. On the other hand, there is no assertion macro which cannot be deactivated via compile time switch. I think it makes more sense to have an `IOX_REQUIRE` which acts like an `IOX_ASSUME` but cannot turned off by a compile time switch.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Relates to #1032

### CODE REVIEW

**Review by elfenpiff — DISMISSED** [review]

**Review by elfenpiff — APPROVED** [review]

File: doc/design/error_reporting.md [hunk 1]
```diff
@@ -165,7 +165,7 @@ the case that it does not hold
 ```cpp
 int x;
 // ...
-IOX_REQUIRE(x>=0, Code::OutOfBounds);
+IOX_REQUIRE(x>=0, "required condition violation message");
```
Comment (MatthiasKillat) [inline doc/design/error_reporting.md:168]: @elBoberido I know it is merged but I think it is useful to have a code or more generic error object if it is to be passed to another framework (this is still the intention).

Some thing like

```IOX_REQUIRE(x>=0, Foo("Text", 73))```.

With sufficient support in the handler, anything can be passed and handled. In particular string literals, runtime constructed strings, objects of any kind, error wrappers around these. As of now, there exist no handling code for anything but codes (or string literals now). We decided this for simplicity.

It is important not to require to pass heavyweight objects, pointers to string literals or error codes are fine, but still allow more heavyweight objects if the user desires. Messages alone may not be descriptive enough at the handler site.

I would also liked to have the message (or any arguments) as optional, but variadic arguments do not play nice in some circumstances, but mainly if the case of no arguments must be supported (not necessarily the case here).

The semantics are also important. `IOX_REQUIRE` is not necessarily on a rare path in the "should not happen category (this would be `IOX_ASSUME`). Therefore performance is a consideration.

It may be a good idea to rethink whether we want more generic objects to be handled in the future. This should include string literals, codes and anything else. Fine for now and can be extended if we see the need.

File: iceoryx_hoofs/reporting/include/iox/assertions.hpp [hunk 2]
```diff
@@ -44,19 +44,29 @@
...
+    do                                                                                                                 \
+    {                                                                                                                  \
+        if (!(expr))                                                                                                   \
+        {                                                                                                              \
+            iox::er::forwardFatalError(iox::er::Violation::createRequiredConditionViolation(),                         \
+                                       iox::er::REQUIRED_CONDITION_VIOLATION,                                          \
+                                       CURRENT_SOURCE_LOCATION,                                                        \
+                                       message); /* @todo iox-#1032 add strigified 'expr' as '#expr' */                \
+        }                                                                                                              \
+    } while (false)
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/reporting/include/iox/assertions.hpp:68]: @elBoberido I recently considered as trailing `static_assert` as an alternative to this `do { ... } while` madness: https://stackoverflow.com/questions/35530850/how-to-require-a-semicolon-after-a-macro.

However, it is not equivalent in general as it e.g. does not play nice in `if(...) MACRO() else { }` without parentheses. One could argue this is a good thing but I disagree as the MACRO should not impose these restrictions if possible. Also there may be more limitations. 

In other words I have no better alternative yet and this is pretty standard but also ancient from the C world.

Stringification as in the TODO is a good addition.

File: doc/design/error_reporting.md (hunk 1 above)
Comment (elBoberido) [reply doc/design/error_reporting.md:168]: @MatthiasKillat for that there is still `IOX_REPORT_IF` and semantically it fits also better if it is intended to be forwarded. Currently I'm even thinking about simplifying the macros even more. Instead of having `IOX_PRECONDITION`, `IOX_ASSUME` and `IOX_REQUIRE` it could be `IOX_ASSERT` and `IOX_ENFORCE`. The former would be deactivated for release builds by default and nicely match the C assert. The latter would would emphasis the stronger requirements of always being active. With this the `IOX_REQUIRE` would again become available if we want to have it as a shortcut for `IOX_REPORT_IF`. We can discuss this further in the currently active PR.
